userapi/api.go

```diff
@@ -0,0 +1,278 @@
+//
+//Unless required by applicable law or agreed to in writing, software
+//distributed under the License is distributed on an "AS IS" BASIS,
+//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//See the License for the specific language governing permissions and
+//limitations under the License.
+
+package userapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+)
+
+const (
+	httpTimeout = 60 * time.Second
+	timeout     = 30 * time.Second
+	apiBasePath = "/api/v1"
+	authUserURL = "/authenticate_user"
+	TestAPIHost = "api-test.cacophony.org.nz"
+	ShortTTL    = "short"
+	MediumTTL   = "medium"
+	LongTTL     = "long"
+)
+
+type CacophonyUserAPI struct {
+	username      string
+	httpClient    *http.Client
+	serverURL     string
+	token         string
+	authenticated bool
+}
+
+// joinURL creates an absolute url with supplied baseURL, and all paths
+func joinURL(baseURL string, paths ...string) string {
+
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		return ""
+	}
+	url := path.Join(paths...)
+	u.Path = path.Join(u.Path, url)
+	return u.String()
+}
+
+func New(conf *Config) *CacophonyUserAPI {
+	api := &CacophonyUserAPI{
+		token:      conf.token,
+		serverURL:  conf.ServerURL,
+		username:   conf.UserName,
+		httpClient: newHTTPClient(),
+	}
+	return api
+}
+
+func (api *CacophonyUserAPI) ServerURL() string {
+	return api.serverURL
+}
+func (api *CacophonyUserAPI) authURL() string {
+	return joinURL(api.serverURL, authUserURL)
+
+}
+
+type Device struct {
+	GroupName  string `json:"groupname"`
+	DeviceName string `json:"devicename"`
+	SaltId     int    `json:"saltId"`
+}
+type DeviceReponse struct {
+	Messages   []string `json:"messages"`
+	Devices    []Device `json:"devices"`
+	StatusCode int      `json:"statusCode"`
+}
+
+func (api *CacophonyUserAPI) User() string {
+	return api.username
+}
+func (api *CacophonyUserAPI) HasToken() bool {
+	return api.token != ""
+}
+func (api *CacophonyUserAPI) IsAuthenticated() bool {
+	return api.authenticated
+}
+
+type tokenResponse struct {
+	Messages []string
+	Token    string
+	ID       int
+}
+
+func (api *CacophonyUserAPI) Authenticate(password string) error {
+	if password == "" {
+		return errors.New("empty password")
+	}
+
+	data := map[string]interface{}{
+		"username": api.username,
+		"password": password,
+	}
+
+	payload, err := json.Marshal(data)
+
+	if err != nil {
+		return err
+	}
+	postResp, err := api.httpClient.Post(
+		api.authURL(),
+		"application/json",
+		bytes.NewReader(payload),
+	)
+	if err != nil {
+		return err
+	}
+	defer postResp.Body.Close()
+
+	if err := handleHTTPResponse(postResp); err != nil {
+		return err
+	}
+
+	var resp tokenResponse
+	d := json.NewDecoder(postResp.Body)
+	if err := d.Decode(&resp); err != nil {
+		return fmt.Errorf("decode: %v", err)
+	}
+	api.token = resp.Token
+	api.authenticated = true
+	if err != nil {
```

Comment:
Error check before writing to `api.token` and `api.authenticated`

Reply:
shouldn't be there removed now